Feat/talk2scholars docker update

ansh-info opened the pull request:
# For authors

## Description
This pull request introduces Docker support for the `talk2scholars` module, aligning it with the existing `talk2biomodels` setup. The changes include:  

1. **Added a Dockerfile for `talk2scholars`**  
   - Uses `python:3.12-slim` as the base image  
   - Installs required dependencies, including `g++` and `build-essential` for compiling `pcst_fast`  
   - Exposes port `8501` for Streamlit  
   - Configures environment variables to be passed at runtime for security  

2. **Updated GitHub Actions (`ci.yml`) to support `talk2scholars`**  
   - Modified workflow to build and push both `talk2biomodels` and `talk2scholars` using a matrix strategy  
   - Ensured environment variables are securely handled via GitHub Secrets  
   - Removed unnecessary test steps for now  

3. **Updated `README.md`**  
   - Documented setup instructions for both `talk2biomodels` and `talk2scholars`  
   - Clarified how users can run containers persistent mode
   - Provided example commands for pulling and running the images  

#### **Context and Motivation**  
The goal of this update is to streamline the deployment process for both `talk2biomodels` and `talk2scholars`, ensuring a consistent and scalable setup. By integrating `talk2scholars` into the CI/CD pipeline, the module can now be built, and deployed automatically alongside `talk2biomodels`.  

This change improves maintainability, security, and usability by:
- Standardizing deployment across all modules  
- Preventing API key exposure by passing them at runtime instead of storing them in the Dockerfile  
- Automating the build and deployment process through GitHub Actions  

#### **Dependencies**  
- Docker  
- GitHub Actions  
- Environment variables (`OPENAI_API_KEY`, `ZOTERO_API_KEY`, `ZOTERO_USER_ID`, `NVIDIA_API_KEY`) must be set when running the container.

## Fixes # (issue) Mention the issue number.

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [ ] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [ ] My code follows the style guidelines mentioned in the Code/DevOps guides
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (e.g. MkDocs)
- [ ] My changes generate no new warnings
- [ ] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

gurdeep330 commented on .github/workflows/ci.yml:
Would it be possible to:
1. Have this workflow automatically run after the RELEASE workflow was successful.
2. Separate it into different jobs based on the agent, and invoke the jobs corresponding to the changes made. For example, if the changes were made to T2S, then only the docker job of T2S is invoked.
3. Automatically increment the version (and not use v1 always) :-)

ansh-info replied:
Hi, Thank you for the review -> Yes we can add these changes, but I want your review first
Please review the below, also regarding the last point we can use GitHub tags, please let me know if this solution will work about tags

### **1. Trigger This Workflow After the `RELEASE` Workflow Completes**  
**Solution:** Modify the workflow trigger to run **after `release.yml` completes successfully** using `workflow_run`:  

**Update `ci.yml`:**  
```yaml
on:
  workflow_run:
    workflows: ["RELEASE"]
    types:
      - completed
  workflow_dispatch:
```

### **2. Separate Jobs by Agent and Run Only for Relevant Changes**  
**Solution:**  
Modify the workflow to **conditionally run jobs** based on changes in `talk2biomodels` or `talk2scholars`.  

**Update `ci.yml`:**  
```yaml
jobs:
  detect-changes:
    runs-on: ubuntu-latest
    outputs:
      build_t2b: ${{ steps.filter.outputs.talk2biomodels }}
      build_t2s: ${{ steps.filter.outputs.talk2scholars }}
    steps:
      - name: Checkout Repository
        uses: actions/checkout@v4

      - name: Detect Changes
        id: filter
        uses: dorny/paths-filter@v2
        with:
          filters: |
            talk2biomodels:
              - 'aiagents4pharma/talk2biomodels/**'
            talk2scholars:
              - 'aiagents4pharma/talk2scholars/**'

  docker-build-t2b:
    needs: detect-changes
    if: needs.detect-changes.outputs.build_t2b == 'true'
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repository
        uses: actions/checkout@v4

      - name: Login to Docker Hub
        uses: docker/login-action@v3
        with:
          username: ${{ vars.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_TOKEN }}

      - name: Build and Push Talk2Biomodels
        uses: docker/build-push-action@v6
        with:
          file: aiagents4pharma/talk2biomodels/Dockerfile
          push: true
          tags: virtualpatientengine/talk2biomodels:latest

  docker-build-t2s:
    needs: detect-changes
    if: needs.detect-changes.outputs.build_t2s == 'true'
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repository
        uses: actions/checkout@v4

      - name: Login to Docker Hub
        uses: docker/login-action@v3
        with:
          username: ${{ vars.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_TOKEN }}

      - name: Build and Push Talk2Scholars
        uses: docker/build-push-action@v6
        with:
          file: aiagents4pharma/talk2scholars/Dockerfile
          push: true
          tags: virtualpatientengine/talk2scholars:latest
```

---

### **3. Automatically Increment the Version Instead of Always Using `v1`**  
**Solution:**  
Modify the workflow to **auto-increment the version** using the latest GitHub tag.  

**Update the `Build and Push` steps in both jobs:**  
```yaml
      - name: Get Version
        id: version
        run: echo "VERSION=$(git describe --tags --abbrev=0)" >> $GITHUB_ENV

      - name: Build and Push
        uses: docker/build-push-action@v6
        with:
          file: aiagents4pharma/talk2biomodels/Dockerfile
          push: true
          tags: |
            virtualpatientengine/talk2biomodels:${{ env.VERSION }}
            virtualpatientengine/talk2biomodels:latest
```

- Uses the **latest Git tag as the version number** instead of hardcoding `v1`.

gurdeep330 replied:
Yes, this looks fine to me. But could you double-check these tests for different conditions by running them in your forked repo? Thx

gurdeep330 reviewed (commented):
Thx @ansh-info for setting this up. Looks very good 👍 

I don't have any comments but I have a request in the workflow file. Let me know if it (or a subset) would be possible in this PR? Thx

ansh-info commented:
HI @gurdeep330 please check the screenshot, The first point you mentioned is completed
"1. Have this workflow automatically run after the RELEASE workflow was successful."
<img width="1419" alt="image" src="https://github.com/user-attachments/assets/2f64cc72-2938-4341-aa72-f65565b8b9fb" />

Let me explain what's happening:

- The workflow is correctly executing and successfully building the Docker image
- It correctly determined the version tag as "v1.10.2" from your repository
- It's correctly attempting to push to Docker Hub using the account "virtualpatientengine"

The error "denied: requested access to the resource is denied" means that my GitHub Actions workflow doesn't have permission to push to that Docker Hub repository. This is expected in a fork scenario

ansh-info commented:
@gurdeep330 Please check the screenshot, I am able to ache the second point you mentioned as well:
"2. Separate it into different jobs based on the agent, and invoke the jobs corresponding to the changes made. For example, if the changes were made to T2S, then only the docker job of T2S is invoked."
<img width="1508" alt="image" src="https://github.com/user-attachments/assets/726df32c-5e95-43c5-9f16-fd038530f08c" />


Previously, with the matrix strategy, when the first job failed (talk2biomodels), the second job (talk2scholars) was automatically canceled. This was because the matrix jobs were part of a single logical job.
Now, with our changes:

- Both modules are running as separate, independent jobs
- Both jobs ran to completion (even though both failed at the same point)
- Each job has its own build summary and record
- Both jobs failed with the same error message about Docker Hub access permissions

This confirms that:

- The path filtering for detecting changes is working (both modules were detected as changed)
- The separation of jobs is working correctly (they run independently)
- The versioning is being properly applied to both jobs

The only error happening now is still the Docker Hub permission issue, which is expected in my fork scenario and doesn't indicate any problem with the workflow structure itself.

ansh-info commented:
@gurdeep330 Your 3rd point is also completed, Please check the attached screenshots
"3. Automatically increment the version (and not use v1 always) :-)"

##### We have a version check workflow added now
<img width="1428" alt="image" src="https://github.com/user-attachments/assets/3a66da48-93af-4b23-a34f-c7fc10c3836f" />

#### Checks for the latest tag from RELEASE workflow
<img width="770" alt="image" src="https://github.com/user-attachments/assets/5c25c988-b667-43a0-a119-5d9a12315d37" />

##### Build summary -> both got build 
<img width="1520" alt="image" src="https://github.com/user-attachments/assets/88a7f499-87f6-4456-bda3-eeb3e07ebd52" />

Implementation of step 3: automatically incrementing the version instead of using a static version. Since our project is already using semantic-release in the RELEASE workflow, we leveraged that work and ensure the Docker images use the correctly incremented versions.

I've implemented step 3, adding automatic version incrementing with a sophisticated approach:

### Key Improvements:

1. **Centralized Version Generation**:
   - Added a new `version` job that computes version information once
   - Other jobs depend on this job and use its outputs
   - This ensures consistent versioning across all Docker images

2. **Smart Version Types**:
   - For releases triggered by the RELEASE workflow, uses the proper semantic version from git tags
   - For development builds (PRs, manual runs), creates a development version like `v1.2.3-dev.5` where 5 is the number of commits since the last tag
   - Also generates a short Git SHA for an additional tag option

3. **Multiple Docker Tags**:
   - Each image now gets tagged with:
     - The specific version number (`v1.2.3` or `v1.2.3-dev.5`)
     - The Git short SHA for easy traceability (`a1b2c3d`)
     - The `latest` tag for convenience

This approach gives you several benefits:
- Images built from release workflows get clean semantic version tags
- Images built from PRs or development builds get clearly marked development version tags
- Using the Git SHA as a tag makes it easy to trace builds back to specific commits
- The version calculation is centralized, ensuring consistent versioning across modules

This implementation completes the third and final step requested in the PR review. The workflow now:
1. Runs after the RELEASE workflow
2. Separates jobs based on which modules have changed
3. Automatically increments version numbers appropriately

Let me know if you'd like any further refinements to this implementation!

ansh-info commented:
Hello @gurdeep330, Just to make sure, I change the username to my GitHub credentials on my test fork, and everything ran successfully. Please check the screenshots bellow(-> This for just to test I have removed it from my docker hub😊):

#### Github Actions Overview
<img width="1785" alt="image" src="https://github.com/user-attachments/assets/bbec83c7-bc94-4b1d-8ded-9eaa2bfb2f67" />


#### Docker Hub preview
<img width="1770" alt="image" src="https://github.com/user-attachments/assets/db99062a-b7c7-4782-ac34-a4767e1b659c" />


<img width="1785" alt="image" src="https://github.com/user-attachments/assets/9ba7bca6-5aa3-4b70-8400-662fb1e10b9e" />

gurdeep330 commented:
Thanks very much @ansh-info This is very nicely done 💯 I'll run a few tests after the merge is complete. Great work 🥇 

You just need an approval from @dmccloskey :-)

dmccloskey reviewed (approved):
Very nice work with the CI and DockerHub build and push 👍.
